### PR TITLE
Data stream stats: Add name only if present

### LINKED ISSUE #337: [BUG] Querying Data Streams Stats with AWS Auth without a name returns a signing error
### What is the bug?
When querying the Data Stream Stats API with AWS Authentication and without setting an index name the AWS Signature is invalid.

Error:
```
The request signature we calculated does not match the signature you provided. 
Check your AWS Secret Access Key and signing method.
```

### How can one reproduce the bug?
Setup an AWS Signer and OpenSearch Client and make a request for `/_data_streams/_stats` without setting an index name.
```go
awsCfg, err := config.LoadDefaultConfig(context.Background())
if err != nil {
    return err
}

awsSigner, err := awsv2.NewSigner(awsCfg)
if err != nil {
    return err
}

url := "https://my-opensearch-cluster"
osCfg := opensearch.Config{
    Addresses: []string{url},
    Signer:    awsSigner,
}

client, err := opensearch.NewClient(osCfg)
if err != nil {
    return err
}

resp, err := client.Indices.GetDataStreamStats()
if err != nil {
    return err
}
```

### What is the expected behavior?
The list of data streams and their stats is returned. The OpenSearch API supports making this request with no name and should return all of the data streams instead.

### What is your host/environment?
Linux 6.2.6
Go 1.20.5

### Do you have any screenshots?
No

### Do you have any additional context?
It looks like the GetDataStreams function always adds the name and thus ends up adding an additional `/` to the path. This `/` is dropped on the server side but does get factored into the signature calculation on the client side causing the mismatch.

### PR DESCRIPTION
**Pull request by DewaldV**
### Description
This makes the GetDataStreamStats request only add the `Name` and path separator (`/`) to the path if Name is not empty.

As it stands when using this function without setting a name the path will be:
```
/_data_stream//_stats
```

This path is then used by the AWS Signer to calculate a signature. When the request is sent the server recalculates the signature but the additional `/` has been dropped by now and we get a signature mismatch. See #337 for details.

I have tested this fix in our own codebase and it resolves the issue as described.

### Issues Resolved
Closes #337 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dblock — CHANGES_REQUESTED** [review]
Looks good, care to add a test, please?

**Review by VachaShah — COMMENTED** [review]
Thank you @DewaldV! +1 for adding a test for this.

**Comment by DewaldV** [comment]
Thanks for taking a look, I'll add a test to this as requested. :+1: 

Apologies for my delayed response I've been at conference this week.

**Comment by DewaldV** [comment]
@VachaShah @dblock apologies for the delay in getting those tests sorted but I've added them as requested.

The integration test is a bit strange since I had to add additional Data Stream create and delete tests so that I could ensure I was actually getting multiple results from the `/_data_stream/_stats` endpoint as expected. Not sure if you're prefer we split it out to a different test so let me know.

I added a unit test to specifically check the behaviour of setting the `Name` field of the `IndicesGetDataStreamStatsRequest` as that was causing my original issue. I felt the Integration Test didn't really cover that specifically so a unit test would be a good way to ensure the URL is valid.

**Review by dblock — DISMISSED** [review]
This is perfect, my favorite kind of PR that has way more tests than an otherwise trivial fix!

**Comment by dblock** [comment]
@DewaldV see failures in tests + missing license header

**Comment by DewaldV** [comment]
@dblock Apologies that I missed that. Just pushed the license headers!

**Comment by dblock** [comment]
> @dblock Apologies that I missed that. Just pushed the license headers!

Cool. I disabled the PR behavior to require approvals from maintainers to run for 1st time contributors (made it for new GitHub users only)

**Comment by DewaldV** [comment]
Ahh, the test failure I thought might be an issue but I ran a it a few times locally and it seemed OK so I left it as is.

I suspect the `/_data_stream/_stats` endpoint returns the data streams in any order.

When I tested locally it seemed consistent but it's likely consistent for my local cluster only compared to running it somewhere else.

I'll have to refactor that test to change how we do the assertion so its not order dependent.

**Comment by DewaldV** [comment]
Hey @dblock. :wave:

After some research and attempts at fixing the ordering issue in tests I found the [jsonassert](https://github.com/kinbiko/jsonassert) library that tackles exactly these sorts of issues. It might be a very useful tool for other tests in this lib given the usage of JSON for all OpenSearch APIs.

It includes functionality for asserting JSON objects match without considering array order and should fix the flakiness in my original test.

I understand introducing a new dependency isn't ideal and might need more conversation but this is only used in Integration Tests at this point so hopefully it can be considered.

**Comment by DewaldV** [comment]
I've checked the failures on the latest CI run and I can't replicate them locally. 

I'm convinced they are unrelated to my changes and tests and may just need a rerun to pass.

I can do some more digging if needed thoug, let me know what you think. :+1:

**Review by dblock — APPROVED** [review]

**Pull request merged**
